From 1bc68c1e9ee8e24806ee82d34ed0e5755f3aca13 Mon Sep 17 00:00:00 2001
From: Virendra Pathak <vpathak@broadcom.com>
Date: Thu, 24 Oct 2013 23:09:58 +0530
Subject: fmn: added popq class range check

- popq priotity class should not be configured if class is out of range
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Nam Ninh <nam.ninh@windriver.com>

diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h
index 4f6ce14..e072903 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn.h
@@ -56,6 +56,7 @@
 
 #define POPQ_CLASS_CONFIG_BASE  	0x4068
 #define POPQ_CLASS_CONFIG_OFFSET	8
+#define POPQ_MAX_CLASS_NUM		3	/* popq priority class: 0, 1, 2, 3 */
 
 enum FMN_MSG_BLKS {
 	XLP_MSG_BLK_CPU = 0, 
diff --git a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
index c818dcf..7492dec 100644
--- a/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
+++ b/arch/mips/netlogic/lib/fmnlib/nlm_hal_fmn_config.c
@@ -637,6 +637,12 @@ int brcm_hal_config_popq_class(int node, int popq_class, uint8_t width, uint16_t
 	int offset = 0;
 	
 	if (is_nlm_xlp9xx()) {
+		if ((popq_class < 0) || (popq_class > POPQ_MAX_CLASS_NUM)) {
+			nlm_print("ERROR: popq class number out of range. \
+					popq_class: %d, popq class range: 0 - %d\n", \
+					popq_class, POPQ_MAX_CLASS_NUM);
+			return -1;
+		}
 		offset = popq_class * POPQ_CLASS_CONFIG_OFFSET;
 		val |= ((1ULL << 63) | ((0x7f & (~width)) << 16) | (base_id & 0x1fff));
 		nlh_write_cfg_reg64(xlp_fmn_base[node] + (POPQ_CLASS_CONFIG_BASE + offset), val);
-- 
1.7.1

